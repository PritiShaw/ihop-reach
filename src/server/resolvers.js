import { MongoClient } from 'mongodb'
//Connection to MongoDB Database
const context = () =>
  MongoClient.connect('mongodb://localhost:27017', {
    useNewUrlParser: true,
  }).then(client => client)
const collection = 'articles' // Set collection name
const dbName = 'iHOP'

const resolvers = {
  //	It returns all Document(250 per page) present in the database
  allDocuments: (args, context) =>
    context().then(async client => {
      let db = client.db(dbName)
      const fieldArr = {
        _id: '_id',
        hypothesis_information: 'extracted_information.hypothesis_information',
        interaction_type: 'extracted_information.interaction_type',
        negative_information: 'extracted_information.negative_information',
        pmc_id: 'pmc_id',
        trigger: 'trigger',
        evidence: 'evidence',
        Species: 'extracted_information.context.Species',
        // entity_text : "extracted_information.participant_a/b.entity_text",
        // entity_type : "extracted_information.participant_a/b.entity_type",
        // identifier : "extracted_information.participant_a/b.identifier"
      }
      let andArray = []
      let orArray = []
      let regx, tempRes
      for (let arg in args) {
        regx = new RegExp(args[arg], 'i')
        switch (arg) {
          case 'page':
            break
          case 'entity_type':
            orArray = []
            orArray.push(
              {
                'extracted_information.participant_a.entity_type': regx,
              },
              {
                'extracted_information.participant_b.entity_type': regx,
              }
            )
            andArray.push({
              $or: orArray,
            })
            break
          case 'entity_text':
            orArray = []
            orArray.push(
              {
                'extracted_information.participant_a.entity_text': regx,
              },
              {
                'extracted_information.participant_b.entity_text': regx,
              }
            )
            andArray.push({
              $or: orArray,
            })
            break
          case 'identifier':
            orArray = []
            orArray.push(
              {
                'extracted_information.participant_a.identifier': regx,
              },
              {
                'extracted_information.participant_b.identifier': regx,
              }
            )
            andArray.push({
              $or: orArray,
            })
            break
          default:
            let fieldObj = {}
            if (
              arg == 'negative_information' ||
              arg == 'hypothesis_information'
            )
              fieldObj[fieldArr[arg]] = args[arg]
            else if (arg == '_id') {
              args.page = 0
              fieldObj['_id'] = new ObjectId(args._id)
            } else fieldObj[fieldArr[arg]] = regx
            andArray.push(fieldObj)
            break
        }
      }
      let filter = {}
      if (andArray.length > 0) filter['$and'] = andArray
      let res
      res = await db
        .collection(collection)
        .aggregate(
          [
            { $match: filter },
            { $skip: args.page < 1 ? 0 : (args.page - 1) * 250 },
            { $limit: 250 },
            //   {$lookup: {
            //     'from': 'pubmed',
            //     'let': {
            //         'pmc': '$pmc_id'
            //     },
            //     'pipeline': [
            //         {
            //         '$match': {
            //             '$expr': {
            //                 '$or' : [
            //                     {'$eq': ['$pmcid', {'$concat' : ["PMC",'$$pmc']}]},
            //                     {'$eq': ['$pmcid', '$$pmc']}
            //                 ]
            //             }
            //         }
            //         }
            //     ],
            //     'as': 'pubmed'
            //     }
            // }, {
            //     '$addFields': {
            //     'publication_year': {
            //         '$ifNull': [
            //         {
            //             '$arrayElemAt': ['$pubmed.year', 0]
            //         },  ''
            //         ]
            //     }
            //     }
            // }, {
            //     '$sort': {
            //         'publication_year': -1
            //     }
            // }
          ],
          {
            allowDiskUse: true,
          }
        )
        .toArray()
      client.close()
      return res
    }),
  // It returns PubMed details by PMCID
  getPubMedDetails: (args, context) =>
    context().then(async client => {
      let db = client.db(dbName)
      let query = args.pmcid.toUpperCase()
      // let res = await db.collection("pubmed").find({"pmcid" : query}).limit(1)
      let res = await db
        .collection('pubmed')
        .findOne({ $or: [{ pmcid: query }, { pmcid: 'PMC' + query }] })
      client.close()
      return res
    }),
  // Returns all the document of the specified identifier
  documentsByIdentifier: (args, context) =>
    context().then(client => {
      let db = client.db(dbName)
      const id = args.identifier.trim()
      const regx = id
      let nameArr = []
      return db
        .collection(collection)
        .aggregate(
          [
            {
              $match: {
                $or: [
                  {
                    'extracted_information.participant_a.identifier': regx,
                  },
                  {
                    'extracted_information.participant_b.identifier': regx,
                  },
                ],
              },
            },
            {
              $lookup: {
                from: 'pubmed',
                let: {
                  pmc: '$pmc_id',
                },
                pipeline: [
                  {
                    $match: {
                      $expr: {
                        $or: [
                          { $eq: ['$pmcid', { $concat: ['PMC', '$$pmc'] }] },
                          { $eq: ['$pmcid', '$$pmc'] },
                        ],
                      },
                    },
                  },
                ],
                as: 'pubmed',
              },
            },
            {
              $addFields: {
                publication_year: {
                  $ifNull: [
                    {
                      $arrayElemAt: ['$pubmed.year', 0],
                    },
                    '',
                  ],
                },
                journal_title: {
                  $ifNull: [
                    {
                      $arrayElemAt: ['$pubmed.journal_title', 0],
                    },
                    '',
                  ],
                },
              },
            },
            {
              $sort: {
                publication_year: -1,
              },
            },
          ],
          {
            allowDiskUse: true,
          }
        )
        .toArray()
        .then(arr => {
          client.close()
          return {
            documents: arr,
            count: arr.length,
            searchkey: id,
          }
        })
    }),
  //	It returns details of all Entities present in database by identifiers.
  allIdentifiers: (args, context) =>
    context().then(async client => {
      let db = client.db(dbName)
      let res = await db
        .collection('identifier_mapping')
        .find()
        .toArray()
      client.close()
      let limit = args['limit'] > 0 ? args['limit'] : res.length
      let x, y
      return res
        .sort((a, b) => {
          x = a.iden
          y = b.iden
          if (x < y) return -1
          else if (x > y) return 1
          else return 0
        })
        .slice(0, limit)
    }),
  //	It returns Entity details by identifier
  identifier: (args, context) =>
    context().then(async client => {
      let db = client.db(dbName)
      let res = await db
        .collection('identifier_mapping')
        .findOne({ iden: args.identifier })
      client.close()
      return res
    }),
}

export { resolvers, context }
