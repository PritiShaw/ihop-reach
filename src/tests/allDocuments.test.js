import graphqlHTTP from 'express-graphql'
import {resolvers,context} from '../resolvers'
import graphQL_Schema from '../schema.graphql'
import {buildSchema} from 'graphql'

const schema =  buildSchema(graphQL_Schema)

const middleware = graphqlHTTP({
    schema,
    rootValue: resolvers,
    context
})

const request = (met,q) =>  ({
    method: met,
    headers: {},
    //replace with your query
    body: { query: q },
})

const response = {
    setHeader: jest.fn(),
    end: jest.fn(),
    json: jest.fn(),
}
it('Test POST allDocuments Query', async () => {
    jest.setTimeout(30000);
    let method = 'POST'
    let query = `{allDocuments(identifier: "chebi:CHEBI:10006"){pmc_id extracted_information {participant_b{identifier}participant_a{identifier}}}}`
    await middleware(request(method,query), response)
    const responseData = response.json.mock.calls[0][0]
    expect(responseData).toMatchSnapshot()
})