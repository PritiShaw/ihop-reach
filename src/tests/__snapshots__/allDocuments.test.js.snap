// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test POST allDocuments Query 1`] = `
Object {
  "data": Object {
    "allDocuments": Array [
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:10008",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
        },
        "pmc_id": "3077501",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [],
          "participant_b": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
        },
        "pmc_id": "3725501",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [],
          "participant_b": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
        },
        "pmc_id": "3705330",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:3658",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
        },
        "pmc_id": "3685533",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "chebi:CHEBI:22984",
            },
          ],
        },
        "pmc_id": "3558159",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "chebi:CHEBI:26078",
            },
          ],
        },
        "pmc_id": "3558159",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "interpro:IPR003625",
            },
          ],
        },
        "pmc_id": "3427190",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "uazid:UAZ147281",
            },
          ],
        },
        "pmc_id": "2850091",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "uniprot:Q07973",
            },
          ],
        },
        "pmc_id": "2850091",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "chebi:CHEBI:24706",
            },
          ],
        },
        "pmc_id": "3376220",
      },
      Object {
        "extracted_information": Object {
          "participant_a": Array [
            Object {
              "identifier": "chebi:CHEBI:10007",
            },
          ],
          "participant_b": Array [
            Object {
              "identifier": "interpro:IPR000324",
            },
          ],
        },
        "pmc_id": "3725501",
      },
    ],
  },
}
`;
